phpBB/phpbb/db/tools/doctrine.php

```diff
@@ -0,0 +1,907 @@
+ */
+class doctrine implements tools_interface
+{
+	/**
+	 * @var AbstractSchemaManager
+	 */
+	private $schema_manager;
+
+	/**
+	 * @var Connection
+	 */
+	private $connection;
+
+	/**
+	 * @var bool
+	 */
+	private $return_statements;
+
+	/**
+	 * Database tools constructors.
+	 *
+	 * @param Connection $connection
+	 * @param bool       $return_statements
+	 */
+	public function __construct(Connection $connection, bool $return_statements = false)
+	{
+		$this->return_statements = $return_statements;
+		$this->connection = $connection;
+	}
+
+	/**
+	 * @return AbstractSchemaManager
+	 *
+	 * @throws Exception
+	 */
+	protected function get_schema_manager(): AbstractSchemaManager
+	{
+		if ($this->schema_manager == null)
+		{
+			$this->schema_manager = $this->connection->createSchemaManager();
+		}
+
+		return $this->schema_manager;
+	}
+
+	/**
+	 * @return Schema
+	 *
+	 * @throws Exception
+	 */
+	protected function get_schema(): Schema
+	{
+		return $this->get_schema_manager()->createSchema();
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function sql_list_tables(): array
+	{
+		try
+		{
+			$tables = array_map('strtolower', $this->get_schema_manager()->listTableNames());
+			return array_combine($tables, $tables);
+		}
+		catch (Exception $e)
+		{
+			return [];
+		}
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function sql_table_exists(string $table_name): bool
+	{
+		try
+		{
+			return $this->get_schema_manager()->tablesExist([$table_name]);
+		}
+		catch (Exception $e)
+		{
+			return false;
+		}
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function sql_list_columns(string $table_name): array
+	{
+		try
+		{
+			return $this->get_asset_names($this->get_schema_manager()->listTableColumns($table_name));
+		}
+		catch (Exception $e)
+		{
+			return [];
+		}
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function sql_column_exists(string $table_name, string $column_name): bool
+	{
+		try
+		{
+			return $this->asset_exists($column_name, $this->get_schema_manager()->listTableColumns($table_name));
+		}
+		catch (Exception $e)
+		{
+			return false;
+		}
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function sql_list_index(string $table_name): array
+	{
+		return $this->get_asset_names($this->get_filtered_index_list($table_name, true));
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function sql_index_exists(string $table_name, string $index_name): bool
+	{
+		return $this->asset_exists($index_name, $this->get_filtered_index_list($table_name, true));
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function sql_unique_index_exists(string $table_name, string $index_name): bool
+	{
+		return $this->asset_exists($index_name, $this->get_filtered_index_list($table_name, false));
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function perform_schema_changes(array $schema_changes)
+	{
+		if (empty($schema_changes))
+		{
+			return;
+		}
+
+		return $this->alter_schema(
+			function (Schema $schema) use($schema_changes): void
```

Comment:
Why not split it into an actual class method and pass as callable? Would IMHO be a bit cleaner.